File: 2015/Day_16/Ruby/part_2.rb
require "../../input_parsing/parse_input"

def create_sue_hash(sue_value_list)
  sue_hash = Hash.new
  sue_value_list.each do |aunt|
    important_values = aunt.scan(/Sue (\d*): (\w*): (\d*), (\w*): (\d*), (\w*): (\d*)/)
    sue_hash[important_values[0][0]] = {important_values[0][1]=>important_values[0][2],important_values[0][3]=>important_values[0][4], important_values[0][5]=>important_values[0][6]}
  end
  sue_hash
end


if $PROGRAM_NAME == __FILE__
  sue_list = input_per_line('../input.txt')
  sue_hash = create_sue_hash(sue_list)
  sue_hash.keep_if{|key, value| value["children"] == '3' or value["children"] == nil}
  sue_hash.keep_if{|key, value| value["cats"] == nil or value["cats"].to_i > 7 }
  sue_hash.keep_if{|key, value| value["samoyeds"] == '2' or value["samoyeds"] == nil }
  sue_hash.keep_if{|key, value| value["pomeranians"] == nil or value["pomeranians"].to_i < 3}
  sue_hash.keep_if{|key, value| value["akitas"] == '0' or value["akitas"] == nil }
  sue_hash.keep_if{|key, value| value["vizslas"] == '0' or value["vizslas"] == nil }
  sue_hash.keep_if{|key, value| value["goldfish"] == nil or value["goldfish"].to_i < 5 }
  sue_hash.keep_if{|key, value| value["trees"] == nil or value["trees"].to_i > 3}
  sue_hash.keep_if{|key, value| value["cars"] == '2' or value["cars"] == nil }
  sue_hash.keep_if{|key, value| value["perfumes"] == '3' or value["perfumes"] == nil }
  puts sue_hash
end

File: 2015/Day_16/Python/part_2.py
from sys import path
import re
path.insert(0, '../../input_parsing')
import parse_input


def create_aunt_sue_dictionary(list_of_aunts):
    regex = re.compile(r'Sue (\d*): (\w*): (\d*), (\w*): (\d*), (\w*): (\d*)')
    aunts = {}
    for aunt in list_of_aunts:
        matches = re.findall(regex, aunt)
        aunts[matches[0][0]] = {
            matches[0][1]: int(matches[0][2]),
            matches[0][3]: int(matches[0][4]),
            matches[0][5]: int(matches[0][6]),
        }

    return aunts


def find_the_aunt(aunt_dictionary):
    potential_aunts_2 = aunt_dictionary.copy()
    for key, values in aunt_dictionary.items():
        if values.get("children") not in [3, None]:
            del potential_aunts_2[key]
        elif values.get('cats') is not None and values.get("cats") <= 7:
            del potential_aunts_2[key]
        elif values.get("samoyeds") not in [2, None]:
            del potential_aunts_2[key]
        elif values.get('pomeranians') is not None and values.get("pomeranians") >= 3:
            del potential_aunts_2[key]
        elif values.get("akitas") not in [0, None]:
            del potential_aunts_2[key]
        elif values.get("vizslas") not in [0, None]:
            del potential_aunts_2[key]
        elif values.get('goldfish') is not None and values.get("goldfish") >= 5:
            del potential_aunts_2[key]
        elif values.get('trees') is not None and values.get("trees") <= 3:
            del potential_aunts_2[key]
        elif values.get("cars") not in [2, None]:
            del potential_aunts_2[key]
        elif values.get("perfumes") not in [1, None]:
            del potential_aunts_2[key]
    return potential_aunts_2


if __name__ == "__main__":
    all_the_aunts = parse_input.input_per_line('../input.txt')
    aunt_sue_dictionary = create_aunt_sue_dictionary(all_the_aunts)
    print(f"The Aunt Sue who sent the present is {find_the_aunt(aunt_sue_dictionary)}")
